---
title: "class15"
author: "Nicole Manuguid"
date: "11/16/2021"
output: pdf_document
---

# Background
Our data for today came from Himes. et al. RNASeq analysis of the drug dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Read the countData and colData

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
  
```

Let's have a look at these

```{r}
metadata
```

```{r}
head(counts)

```

>Q1. How many genes are in this dataset? 

```{r}
nrow(counts)
```

>Q2. How many ‘control’ cell lines do we have? 

```{r}
sum(metadata$dex == "control")
```
First I need to extract all the "control" columns. Then I will take the rowwise mean to get the average conunt values for all genes in these four experiments. 

>Q3. How would you make the above code in either approach more robust?

Avoid dividing by 4 because in the future more data could be added to this dataset. 

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[, control.inds]
head(control.counts)
```


```{r}
control.mean <- rowMeans(control.counts)
# control.mean
```

Now do the same for the drug treated experiments (ie. columns)

> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.inds <- metadata$dex =="treated"
treated.counts <- counts[, treated.inds]
head(treated.counts)
treated.mean <- rowMeans(treated.counts)
```


We will combine our meancount data for bookkeeping purposes.

```{r}
meancounts <- data.frame(control.mean, treated.mean)
# meancounts
```

```{r}
colSums(meancounts)
```

Let's make a quick plot

>Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.


```{r}
plot(meancounts)
```

This screams at us that we need a log transformation to see details of our data!

I am going to re-plot on a log scale

```{r}
plot(meancounts, log = "xy")
```

>Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot? 

```{r}
library(ggplot2)
ggplot(meancounts, aes(x=control.mean, y=treated.mean)) +
  geom_point() + 
  scale_x_log10() +
  scale_y_log10()

```

We often use log2 in this field because it have nice math properties that make interpretation easier. 

```{r}
log2(10)
```
```{r}
log2(5/10)
```
```{r}
log2(40/10)
```


Cool we see 0 values for no change and + values for increases and minus values for decreases. This nice property leads us to work with **log2(fold-change)** all the time in the genomics and proteomics filed. 

Let's add the **log2(fold-change)** values to our 'meancounts' dataframe

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/
                            meancounts[,"control.mean"])
head(meancounts)
```

I need to exclude the genes (i.e. rows) with zero counts because we can't say anything about these as we have no data for them!

>Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?


```{r}
head(meancounts[, 1:2])
```

```{r}
head(meancounts[,1:2]==0)
```

```{r}
which( c(F,F,T,T))
```
I can use the **which()** function with the 'are.ind=TRUE' argument to get the columns and rows where the TRUE values are (i.e. the zero counts in our case).

```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind = TRUE)
head(zero.vals)
```


```{r}
to.rm <- unique(zero.vals[,"row"])
head(sort(to.rm))
```

Now remove these from our 'meancounts' dataframe.

```{r}
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

How many do we have left 
```{r}
nrow(mycounts)
```

>Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level? 

```{r}
up.ind <- mycounts$log2fc > 2
sum(up.ind)
```

>Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level? 

```{r}
down.ind <- mycounts$log2fc < (-2)
sum(down.ind)
```

>Q10. Do you trust these results? Why or why not?

We should not trust these results becuase the data was analyzed with a fold change. In order to trust these results, we should determine whether differences in data are statistically significant. 

#DESeq2 analysis

```{r}
library(DESeq2)
citation("DESeq2")
```

We need to first setup the input for DESeq

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

Now we can run DESeq analysis 

results() first produces an error, need to run DESeq first
```{r}
dds <- DESeq(dds)
```
To get at the results here we use the DESeq 'results()' function: 
```{r}
res <- results(dds)
head(res)
```

```{r}
summary(res)
```

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

##Save our results

Write out whole results dataset (including genes that do not change significantly)
```{r}
write.csv(res, file="allmyresults.csv")
```

Focus in on those genes with a small p-value (i.e. show a significant change).

```{r}
res05 <- results(dds, alpha=0.05)
```

```{r}
summary(res05)
```


#data visualization - volcano plot

Let’s make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated and display a high fold change.

```{r}
plot(res$log2FoldChange, res$padj)
```

That is not a useful plot because all the small p-values are hidden at the bottom of the plot and we can't really see them.

Log will help.
```{r}
plot(res$log2FoldChange, log(res$padj))
```

We can flip this p-value axis by just putting a minus sign on it then we will have the classic volcano plot that the rest of the world uses.

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

Finally let's add some color to this plot to draw attention to the genes (i.e. points) we care about - that is those with large fold-change and low p-values (i.e. high -log(p-values))


```{r}
plot(res$log2FoldChange, -log(res$padj))
abline (v=c(-2, +2), col = "gray")
abline(h=-log(0.05), col = "gray")
```

Now we can add some color to the points 

```{r}
#set up color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

#volcano plot
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

#cut off lines 
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```


#Adding Annotation data for our genes

For this we need two bioconductor packages
-BiocManager::install("AnnotationDbi")
-BiocManager::install("org.Hs.eg.db")

```{r}
#Using mapIDs() function from bioconductor 

library("AnnotationDbi")
library("org.Hs.eg.db")
```

Let's have a look at what is in the 'org.Hs.eg.db'
```{r}
columns(org.Hs.eg.db)
```


We will use the 'mapIDs" function to translate between identifiers from different databases.

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",  # The format
                     column="SYMBOL",   # new format we want
                     multiVals="first")
```

We need ENTREZ ids or pathway analysis with KEGG.

```{r}
columns(org.Hs.eg.db)
```

>Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",  # The format
                     column="ENTREZID",   # new format we want
                     multiVals="first")
```

```{r}
head(res)
```
Let's make a another volcano plot with some gene labels. For this we can use the EnhancedVolcano package
```{r}
BiocManager::install("EnhancedVolcano")
```
```{r}
library(EnhancedVolcano)
```


```{r}
x <- as.data.frame(res)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```

#Pathway analysis/geneset annotation 

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

```


```{r}
#look at first two pathways 
head(kegg.sets.hs, 2)
```

The main **gage()** funciton requires a named vector of fold changes, where the names of the values are the Entrz gene IDs
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

This separates out results by "greater" and "less" i.e., those that are up regulated and those that are down regulated.
```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```

Now, let’s try out the pathview() function from the pathview package to make a pathway plot with our RNA-Seq expression results shown in color.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```


![](hsa05310.pathview.png)


>Q12. Can you do the same procedure as above to plot the pathview figures for the top 2 down-reguled pathways?

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa05310", kegg.native=FALSE)
```

